components/tikv_util/src/resizable_threadpool.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::sync::{Arc, RwLock};
+
+use futures::Future;
+use tokio::{io::Result as TokioResult, runtime::Runtime};
+
+/// DaemonRuntime is a "background" runtime, which contains "daemon" tasks:
+/// any task spawn into it would run until finish even the runtime isn't
+/// referenced.
+pub struct DaemonRuntime(Option<Runtime>);
+
+impl DaemonRuntime {
+    /// spawn a daemon task to the runtime.
+    pub fn spawn(self: &Arc<Self>, f: impl Future<Output = ()> + Send + 'static) {
+        let wkr = self.clone();
+        self.0.as_ref().unwrap().spawn(async move {
+            f.await;
+            drop(wkr)
+        });
+    }
+
+    /// create a daemon runtime from some runtime.
+    pub fn from_runtime(rt: Runtime) -> Arc<Self> {
+        Arc::new(Self(Some(rt)))
+    }
+}
+
+impl Drop for DaemonRuntime {
+    fn drop(&mut self) {
+        // it is safe because all tasks should be finished.
+        self.0.take().unwrap().shutdown_background()
+    }
+}
+
+pub struct ResizableRuntime {
+    pub size: usize,
+    thread_name: String,
+    pool: Option<Arc<DaemonRuntime>>,
+    after_adjust: fn(usize),
```

Comment:
Use `Box<Fn(usize)>` instead. `fn(usize)` is a function pointer. Just consider the difference between `std::function<void(int)>` and `void (*fp)(int)`.